_posts/2020-06-26-Making-IrishGen-SPARQL-Part-I.md

```diff
@@ -0,0 +1,549 @@
+present?  As the reader can see from the example above the `optional`
+keyword needs `{` and `}` which implies that more than one thing can
+be optional at a time.  The general case can be re-specified as: the
+subject can have either `irishRel:nomName` or `irishRel:genName` or
+both.  In this case SPARQL has the
+[`union`](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#alternatives)
+keyword which combines the results of graph patterns.  Thus, in the
+case where both `irishRel:nomName` or `irishRel:genName` can be
+optional, the query can be rewritten:
+
+```sparql
+prefix irishRel: <http://example.com/earlyIrishRelationship.ttl#> 
+
+select ?x ?y
+from <tag:stardog:api:context:all>
+where {
+    optional { 
+      {?x irishRel:nomName ?y }
+      union
+      { ?x irishRel:genName ?y } 
+    }
+    filter regex(?y, "^B[áa]eth$", "i")
+}
+```
+
+The above query will return 24 results.  This will not distinguish
+between nominative or genitive and will combine both if found.  Most
+of the queries so far have not taken advantage of the reasoning
+capabilities that were [discussed previously]({% post_url
+2020-06-22-Triplestores-Onotologies-Reasoning %}).  Thus, there is an
+alternative way for this query to be written so that it will do a
+similar search, that is search for names irrespective of their
+grammatical case, but rely instead on reasoning rather than
+optionality.  If the reader consults the definition of `nomName`
+predicate in the
+[`earlyIrishRelationship.ttl`](https://github.com/cyocum/irish-gen/blob/master/earlyIrishRelationship.ttl)
+file, the reader will notice that:
+
+```turtle
+:nomName
+    a owl:DatatypeProperty ;
+    rdfs:subPropertyOf oldIrish:nominative, foaf:name .
+	
+:genName
+    a owl:DatatypeProperty ;
+    rdfs:subPropertyOf oldIrish:genitive, foaf:name .
+```
+
+This means that the nominative name is a sub-property of `foaf:name`.
+In other words, all `:nomName` are `foaf:name`.  Further, `:genName`
+is similarly defined.  Taken as a whole, this means that, in the
+presence of a Triplestore with a sufficiently powerful reasoner,
+`foaf:name` can take the place of the other combination of names.
+Thus:
+
+```sparql
+prefix foaf:  <http://xmlns.com/foaf/0.1/>
+
+select ?x ?y
+from <tag:stardog:api:context:all>
+where {
+    ?x foaf:name ?y
+    filter regex(?y, "^B[áa]eth$", "i")
+}
+```
+
+The result of this query is also 24 as there are very few dative and
+accusative cases encoded in the dataset.
+
+There is a final form of the `select` query that is useful to discuss
+before moving to `construct` queries is
+[aggregates](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#aggregates).
+These kinds of queries generally count or add or generally combines or
+aggregates information, hence the name.  We will need to part ways for
+a moment from our useful idiot, Baeth.  There is encoded in the
+dataset the `:numChild` predicate.  In the genealogies, the number of
```

Comment:
"In the genealogies" > "In the medieval texts of the genealogies" (just to make it clear that this is from the source material)